modules/tinymce/src/themes/silver/main/ts/ui/dialog/IFrame.ts

```diff
@@ -15,40 +16,89 @@ interface IFrameSourcing {
+              scrollToY(win, 'bottom');
+            } else if (!isScrollAtBottom && (isSafari || isFirefox) && lastScrollTop !== 0) {
+              // TINY-10078: Safari and Firefox reset scroll to top on each document.write(), so we need to restore scroll manually
+              scrollToY(win, lastScrollTop);
+            }
+          }
+        };
+
+        // TINY-10078: On Safari, attempting to scroll before iframe has finished loading will cause scroll to reset to top upon load.
+        // We won't do this for all browsers since this does introduce a slight visual lag.
+        if (isSafari) {
+          iframe.addEventListener('load', scrollAfterWrite, { once: true });
+        }
+
+        doc.open();
+        doc.write(html);
+        doc.close();
+
+        if (!isSafari) {
+          scrollAfterWrite();
+        }
+      });
+  };
+
+  // TINY-10097: Throttle to reduce flickering, as the document.write() method still observes significant flickering on Safari.
+  const writeValueThrottler = isSafari ? Optional.some(Throttler.first(writeValue, 500)) : Optional.none();
+
   return {
     getValue: (_frameComponent: AlloyComponent): string =>
       // Ideally we should fetch data from the iframe...innerHtml, this triggers Cors errors
       cachedValue.get(),
     setValue: (frameComponent: AlloyComponent, html: string) => {
-      // TINY-3769: We need to use srcdoc here, instead of src with a data URI, otherwise browsers won't retain the Origin.
-      // See https://bugs.chromium.org/p/chromium/issues/detail?id=58999#c11
       if (cachedValue.get() !== html) {
         const iframeElement = frameComponent.element as SugarElement<HTMLIFrameElement>;
         const setSrcdocValue = () => Attribute.set(iframeElement, 'srcdoc', html);
 
         if (stream) {
-          const iframe = iframeElement.dom;
-          Optional.from(iframe.contentDocument).fold(
-            setSrcdocValue,
-            (doc) => {
-              const isElementScrollAtBottom = ({ scrollTop, scrollHeight, clientHeight }: HTMLElement) => Math.ceil(scrollTop) + clientHeight >= scrollHeight;
-              // TINY-10032: If documentElement is null, we assume document is empty and so scroll is at bottom.
-              const isScrollAtBottom = Optional.from(doc.documentElement).forall(isElementScrollAtBottom);
-
-              doc.open();
-              doc.write(html);
-              doc.close();
-
-              const win = iframe.contentWindow;
-              const body = doc.body;
-              // TINY-10032: Do not attempt to scroll if body has not been loaded yet
-              if (isScrollAtBottom && Type.isNonNullable(win) && Type.isNonNullable(body)) {
-                win.scrollTo(0, body.scrollHeight);
-              }
-            });
+          const args = [ iframeElement, html, setSrcdocValue ] as const;
+          writeValueThrottler.fold(() => writeValue(...args), (throttler) => throttler.throttle(...args));
```

Comment:
I'm not really a fan of using args like this. I'd prefer
```suggestion
          const f = writeValueThrottler.fold(() => writeValue, (throttler) => throttler.throttle);
          f(iframeElement, html, setSrcdocValue)
```

Which works because we never rely on `this`, so `throttler.throttle` can be called without a scope

Reply:
This is a good way to refactor, thanks 👍